controllers/namespace_controller.go

```diff
@@ -33,20 +37,134 @@ type NamespaceReconciler struct {
+	if !ns.DeletionTimestamp.IsZero() {
+		shouldRequeue, err := r.CleanUp(ctx, log, &ns)
+		return ctrl.Result{
+			Requeue: shouldRequeue,
+		}, err
+	}
+
+	if controllerutil.ContainsFinalizer(&ns, danav1alpha1.NsFinalizer) {
+		err := r.Update(ctx, log, &ns)
+		return ctrl.Result{}, err
+	}
+
+	err := r.Init(ctx, log, &ns)
+	return ctrl.Result{}, err
+}
+
+func (r *NamespaceReconciler) Init(ctx context.Context, log logr.Logger, namespace *corev1.Namespace) error {
+	if _, err := ctrl.CreateOrUpdate(ctx, r.Client, namespace, func() error {
+		controllerutil.AddFinalizer(namespace, danav1alpha1.NsFinalizer)
+		return nil
+	}); err != nil {
+		if apierrors.IsConflict(err) {
+			log.Info("newer resource version exists")
+			return nil
+		}
+		log.Error(err, "unable to add finalizer")
+		return err
+	}
+	log.Info("finalizer created")
+	return nil
+}
+
+func (r *NamespaceReconciler) Update(ctx context.Context, log logr.Logger, namespace *corev1.Namespace) error {
+	var snsList danav1alpha1.SubnamespaceList
+	role := danav1alpha1.NoRole
+
+	if err := r.List(ctx, &snsList, client.InNamespace(namespace.Name)); err != nil {
+		return err
+	}
+
+	if len(snsList.Items) == 0 {
+		role = danav1alpha1.Leaf
+	}
+
+	if _, err := ctrl.CreateOrUpdate(ctx, r.Client, namespace, func() error {
+		namespace.Annotations[danav1alpha1.Role] = role
+		return nil
+	}); err != nil {
+		log.Error(err, "unable to update role")
+		return err
+	}
+
+	return nil
+}
+
+func (r *NamespaceReconciler) CleanUp(ctx context.Context, log logr.Logger, namespace *corev1.Namespace) (shouldRequeue bool, err error) {
+	var sns danav1alpha1.Subnamespace
 
-	// your logic here
+	//check if sns is gone
+	if err := r.Get(ctx, client.ObjectKey{
+		Namespace: namespace.Labels[danav1alpha1.Parent],
+		Name:      namespace.Annotations[danav1alpha1.SnsPointer],
+	}, &sns); err != nil {
+		if !apierrors.IsNotFound(err) {
+			log.Error(err, "unable to get Subnamespace")
+			return false, err
+		}
+	} else {
+		//sns still present
+		if err := r.Delete(ctx, &sns); err != nil {
+			log.Error(err, "unable to delete Subnamespace")
+			return false, err
+		}
+		log.Info("sns deleted")
+		return true, nil
+	}
+	//sns is gone did we already remove the finalizer
+	if !controllerutil.ContainsFinalizer(namespace, danav1alpha1.NsFinalizer) {
+		return false, nil
+	}
+	//remove finalizer
+	if _, err := ctrl.CreateOrUpdate(ctx, r.Client, namespace, func() error {
+		controllerutil.RemoveFinalizer(namespace, danav1alpha1.NsFinalizer)
+		return nil
 
-	return ctrl.Result{}, nil
+	}); err != nil {
+		log.Error(err, "unable to remove finalizer")
+		return false, err
+	}
+	log.Info("finalizer removed")
+	return false, nil
 }
 
 func (r *NamespaceReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&corev1.Namespace{}).
+		//filter all namespace without the HNS label
+		WithEventFilter(NamespacePredicate{predicate.NewPredicateFuncs(func(object client.Object) bool {
```

Comment:
Should've declare NamespacePredicate before, otherwise there's no reason for NamespacePredicate struct.

Reply:
I think you are right, but `WithEventFilter` receives an interface of type `predicate.Predicate`, I'm not sure it possible w/o declaring the NamespacePredicate struct.
If you know how to do so, please let me know.